app/controllers/omniauth_callbacks_controller.rb

```diff
@@ -0,0 +1,12 @@
+class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+
+  def facebook
+    if request.env["omniauth.auth"].info.email.blank?
+      redirect_to "/users/auth/facebook?auth_type=rerequest&scope=email,public_profile"
+    else
```

Comment:
@stedmanblake Here's where I think we'd need the most work / updates on this stream of work – the main thing we'd want to handle is the case where a user exists with an email in our DB and we're trying to sync the accounts – you can see the example code in this post here that should give some tips: https://github.com/omniauth/omniauth/wiki/Managing-Multiple-Providers

Don't worry about finding an `Identity`, you can just try to lookup a `User` first with the given email and then do the associating with the FB login (otherwise the `first_or_create` action with the `.from_omniauth` method is going to create a user with the given oauth params, even if the email is the same, so we'd get an error since our `User` model can't have a user with the same exact email address).

As a bonus, we can think about adding some logic for handling how a currently signed in user would associate their account with the FB login (think of a button inside your settings that simply says something like "Connect with Facebook"). If they click on that they should also update their user record with all of the OAuth info so that they can sign in with FB in the future and have their photo associated etc. This would probably require writing a new method on the User called `update_from_omniauth` so as to ensure that we don't override their name and other settings they may have already set when they signed up.